src/compiler/factory.ts

```diff
@@ -0,0 +1,56 @@
+/// <reference path="types.ts"/>
+/// <reference path="utilities.ts"/>
+/* @internal */
+namespace ts {
+    export function createNodeArray<T extends Node>(elements?: T[], location?: TextRange): NodeArray<T>;
+    export function createNodeArray<T extends Node, TArray extends NodeArray<T>>(elements: TArray, location?: TextRange): TArray;
+    export function createNodeArray<T extends Node, TArray extends NodeArray<T>>(elements?: T[], location?: TextRange): TArray {
+        const array = <TArray>(elements || []);
+        if (location) {
+            array.pos = location.pos;
+            array.end = location.end;
+        }
+        else if (array.pos === undefined || array.end === undefined) {
+            array.pos = -1;
+            array.end = -1;
+        }
+
+        return array;
+    }
+
+    export function createNodeArrayNode<T extends Node>(elements?: (T | NodeArrayNode<T>)[]): NodeArrayNode<T> {
+        const array = <NodeArrayNode<T>>createNodeArray(elements);

```

Comment:
It would end up being `createNodeArray<T | NodeArrayNode<T>, NodeArrayNode<T>>(elements))`, which isn't as clear.